testkit/shared/src/main/scala/cats/effect/testkit/TestConsole.scala

```diff
@@ -0,0 +1,569 @@
+              |$logStr
+              |=== Current State ===
+              |$stateStr""".stripMargin
+        }
+
+      override def lowLevel: LowLevel[F] = this
+
+      override def stdOutOperations: F[List[Op.StdOutOp]] = operations.map(_.flatMap {
+        case op: Op.StdOutOp => Chain.one(op)
+        case _ => Chain.empty
+      }.toList)
+
+      override def stdErrOperations: F[List[Op.StdErrOp]] = operations.map(_.flatMap {
+        case op: Op.StdErrOp => Chain.one(op)
+        case _ => Chain.empty
+      }.toList)
+
+      override def stdInOperations: F[List[Op.StdInOp]] = operations.map(_.flatMap {
+        case op: Op.StdInOp => Chain.one(op)
+        case _ => Chain.empty
+      }.toList)
+
+      override def operationsLog: F[List[Op]] = operations.map(_.toList)
+    }
+  }
+
+  sealed trait Op
+  object Op {
+    sealed trait StdOutOp
+    sealed trait StdErrOp
+    sealed trait StdInOp
+
+    final case class Error(value: String) extends Op with StdErrOp
+    final case class Errorln(value: String) extends Op with StdErrOp
+    final case class Print(value: String) extends Op with StdOutOp
+    final case class Println(value: String) extends Op with StdOutOp
+    final case class Write(value: String) extends Op with StdInOp
+    final case class Writeln(value: String) extends Op with StdInOp
+    final case class ReadAttempted(id: Int) extends Op with StdInOp
+    final case class ReadSuccess(id: Int, line: String) extends Op with StdInOp
+    final case class ReadFailure(id: Int, throwable: Throwable) extends Op with StdInOp
+    final case class DiscardStdInContents(lines: Long, bytes: Long) extends Op
+    final case class NotifyPendingReads(requests: Long) extends Op
+    case object Closed extends Op
+
+    implicit val show: Show[Op] = Show.show {
+      case Error(value) => s"error($value)"
+      case Errorln(value) => s"errorln($value)"
+      case Print(value) => s"print($value)"
+      case Println(value) => s"println($value)"
+      case Write(value) => s"Writing to stdIn: $value"
+      case Writeln(value) => s"Writing line to stdIn: $value"
+      case ReadAttempted(id) => s"Reading stdIn [id: $id]"
+      case ReadSuccess(id, line) => s"Read from stdIn [id: $id]: $line"
+      case ReadFailure(id, throwable) => s"Read from stdIn failed [id: $id]: $throwable"
+      case DiscardStdInContents(lines, bytes) =>
+        s"Discarded $lines lines and $bytes bytes from stdIn"
+      case NotifyPendingReads(requests) => s"Notified $requests pending read requests"
+      case Closed => "Closed"
+    }
+  }
+
+  trait TestStdIn[F[_]] {
+
+    /**
+     * Write a string to the simulated stdIn using the system-default charset
+     *
+     * Blocked calls to [[TestConsole.readLineWithCharset]] will be woken up if `str` contains
+     * one or more lines.
+     *
+     * @note
+     *   Blocked calls will be woken in a first-in-first-out order.
+     */
+    def write[A](value: A)(implicit S: Show[A]): F[Unit]
```

Comment:
```suggestion
    def write[A](value: A)(implicit S: Show[A] = Show.fromToString[A]): F[Unit]
```

We try to idiomatically follow this pattern (at least within CE) when dealing with `Show`, since universal `toString` is a better fallback default than compilation failure. All of implicit `Show`s here should be updated accordingly.